### PR TITLE
Create custom backend for Colaraz IdP

### PR DESCRIPTION
**Pull request by HamzaIbnFarooq**
This PR contains a custom Backend compatible with Colaraz's IdP.
We can't use same backend with different sites so, I have made a small change in models.py to skip filtering of backends on the basis of sites.

### CODE REVIEW

File: common/djangoapps/third_party_auth/colaraz.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        current_req = get_current_request()
+
+        environ = getattr(current_req, "environ", {})
+        schema = environ.get("wsgi.url_scheme", "http")
+
+        site = getattr(current_req, "site", None)
+        domain = getattr(site, "domain", None)
+
+        if not domain:
+            raise AuthException("Colaraz", "Error while authentication")
```
Comment (danialmalik) [inline common/djangoapps/third_party_auth/colaraz.py]: Can we use some more descriptive and specific error log here? Mentioning that the error is due to the missing domain.

File: common/djangoapps/third_party_auth/colaraz.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+from social_core.exceptions import AuthException
+from django.utils.functional import cached_property
+
+from openedx.core.djangoapps.theming.helpers import get_current_request
+from third_party_auth.identityserver3 import IdentityServer3
+
```
Comment (mumarkhan999) [inline common/djangoapps/third_party_auth/colaraz.py]: add one blank line.

File: common/djangoapps/third_party_auth/colaraz.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+from social_core.exceptions import AuthException
+from django.utils.functional import cached_property
```
Comment (mumarkhan999) [inline common/djangoapps/third_party_auth/colaraz.py]: sort them.

File: common/djangoapps/third_party_auth/colaraz.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            raise AuthException("Colaraz", "Error while authentication")
+        elif not domain.startswith(user_site_domain.lower()):
+            raise AuthException("Colaraz", "Your account belongs to {}".format(user_site_domain))
+
+        details = {
+            "fullname": "{} {}".format(response["firstName"], response["lastName"]),
+            "email": response["email"],
+            "first_name": response["firstName"],
+            "last_name": response["lastName"],
+            "username": response["email"]
```
Comment (mumarkhan999) [inline common/djangoapps/third_party_auth/colaraz.py]: To remain consistent with edx code base use single quotes where it is possible.

File: common/djangoapps/third_party_auth/colaraz.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+from django.utils.functional import cached_property
+
+from openedx.core.djangoapps.theming.helpers import get_current_request
+from third_party_auth.identityserver3 import IdentityServer3
+
+class ColarazIdentityServer(IdentityServer3):
+    """
+    An extension of the IdentityServer3 for use with Colaraz's IdP service.
+    """
+    name = "colarazIdentityServer"
```
Comment (mumarkhan999) [inline common/djangoapps/third_party_auth/colaraz.py]:
```suggestion
    name = "ColarazIdentityServer"
```

File: common/djangoapps/third_party_auth/colaraz.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        domain = getattr(site, "domain", None)
+
+        if not domain:
+            raise AuthException("Colaraz", "Error while authentication")
+
+        return "{}://{}/auth/complete/{}".format(schema, domain, self.name)
+
+    def get_user_details(self, response):
+        """
+        Return details about the user account from the service
```
Comment (mumarkhan999) [inline common/djangoapps/third_party_auth/colaraz.py]:
```suggestion
        Returns details about the user account from the service
```

File: common/djangoapps/third_party_auth/colaraz.py (hunk 4 above)
Comment (HamzaIbnFarooq) [reply common/djangoapps/third_party_auth/colaraz.py]: Is it really the norm?

File: common/djangoapps/third_party_auth/colaraz.py (hunk 5 above)
Comment (HamzaIbnFarooq) [reply common/djangoapps/third_party_auth/colaraz.py]: It was intentionally made this way, I just mimicked the IdentityServer3 pattern.

File: common/djangoapps/third_party_auth/colaraz.py (hunk 6 above)
Comment (HamzaIbnFarooq) [reply common/djangoapps/third_party_auth/colaraz.py]: done

File: common/djangoapps/third_party_auth/colaraz.py (hunk 3 above)
Comment (HamzaIbnFarooq) [reply common/djangoapps/third_party_auth/colaraz.py]: done

File: common/djangoapps/third_party_auth/colaraz.py (hunk 1 above)
Comment (HamzaIbnFarooq) [reply common/djangoapps/third_party_auth/colaraz.py]: The message is meant to be shown to the user, on his login page. Other than that I have added some logs, will revisit again to make the logs more appropriate.

File: common/djangoapps/third_party_auth/colaraz.py (hunk 2 above)
Comment (HamzaIbnFarooq) [reply common/djangoapps/third_party_auth/colaraz.py]: done

**Comment by tasawernawaz** [comment]
I just saw that you had merged this branch in `colaraz_master` what kind of branch is this? We should follow the same pattern across all projects.

**Pull request merged**